new functionality for review

meesterdude opened the pull request:
I find that using a reflow a lot, i lose track of the PR's it creates, and have a hard time picking up where i left off. 


This makes several changes:
- like git commit, if you do not pass in the message (or title, in this case) it will prompt you with an editor; first line is the title, rest is the body. This is a change of the default behavior of the `review` command, but it's more in line with what I'd expect to happen. 
- if you pass the title and/or the body options to `review`, it will use that and default to the branch/last commit, if you left either off. 
- updates the default title to be the branch name, default body to be the last commit
- prints the PR on the command line to look over
- prompts user before submitting PR; default is yes (enter)
- adds an `-e` flag that allows you to edit in the editor, prefilled with what the default message would be (branch title, last commit body)
- updates README to reflect new workflow
- bumps version to `0.6.0` per major.minor.bugfix methodology

nhance commented on lib/git_reflow/commands/review.rb:
```diff
@@ -5,16 +5,45 @@
+    end
+
+    if editor
+      system('nano', commit_msg_file)
```
Let's use `$EDITOR` if it's set, cause `nano` would make me lose my mind

meesterdude replied:
would love to - but i couldn't find a solution that worked. got an example i can run with?

codenamev replied:
`$EDITOR` is pretty standard, and we're only supporting OS X and Ubuntu for now.  Also, you'll want to take advantage of `GitReflow.run` for running any system commands ([see ref](https://github.com/reenhanced/gitreflow/blob/master/lib/git_reflow/sandbox.rb)).

meesterdude replied:
@codenamev i tried the code you mentioned - it just hangs, can't even ctrl+C, have to kill -9 in another terminal. I think its around sourcing the environment variables. Maybe we default to nano, and allow cfg to set preference? nano is the least controversial as a default

codenamev replied:
ah right, I forgot you need `GitReflow.run("$editor #{commit_msg_file}", with_system: true)` to run it as `system` instead of backticks

nhance commented:
I really like this

codenamev commented on lib/git_reflow/commands/review.rb:
```diff
@@ -5,16 +5,45 @@
 
-    review_options['title'] ||=  GitReflow.get_first_commit_message
-    review_options['body']  ||=  review_options['title']
+    commit_msg_file = '/tmp/git_reflow_pr_msg'
```
I don't think we can guarantee that every system will have a `/tmp` directory.  can we use the `.git` directory and create another file like git does with the commit messages?  maybe `.git/GIT_REFLOW_PR_MSG`

codenamev replied:
the naming here is throwing me too after reading more code below.  let's rename `commit_msg_file` to `pull_request_msg_file`

codenamev commented on lib/git_reflow/commands/review.rb:
```diff
@@ -5,16 +5,45 @@
-    review_options['body']  ||=  review_options['title']
+    commit_msg_file = '/tmp/git_reflow_pr_msg'
 
+    if global_options[:title] || global_options[:message]
```
since `(global_options[:title] || global_options[:message]) == (editor = false)`, let's just assign this above into a variable: `use_editor = ("#{(global_options[:title] || global_options[:message])}".length == 0)`, then this becomes `unless use_editor` and we can eliminate the other assignments of `editor` below

codenamev commented on lib/git_reflow/commands/review.rb:
```diff
@@ -5,16 +5,45 @@
+      pr_msg = File.open(commit_msg_file).each_line.map(&:strip).to_a
+      title = pr_msg.shift
+      File.delete(commit_msg_file)
+      review_options = {'base' => args[0],'title' => title,'body' =>  pr_msg.join("\n")}
```
let's remove surrounding blank lines from the `pr_msg`  I just imagine many users adding an extra return between the title and body.  so `'body' => pr_msg.reject(&:empty?).join("\n")

codenamev commented:
yeah, HUGE :+1: from me on this too :smile_cat: great work!

codenamev commented on lib/git_reflow/commands/review.rb:
```diff
@@ -5,16 +5,45 @@
+    puts "Body:\n#{review_options['body']}\n"
+    puts "--------\n"
+    puts "Submit PR? (Y/n):"
+    exit 0 if STDIN.gets.chop.downcase =~ /n/
```
we're using the highline gem for input: `GitReflow.review review_options if ask("Submit pull request? (y/n)") =~ /y/i`

meesterdude commented:
ok, updated this. 

I addressed most of val's concerns. One of them, the `use_editor` idea i tried but found the code to be less clear, so i undid. 

- I removed the `-e` flag, it'll always provide the branch name as the title, at least as a starting point; i will be confused with a blank editor. 
- simplified some of the code. 
- removed extra space that might appear between title and body
- updated the prompt to be clearer what the default will be with enter key

codenamev commented on lib/git_reflow/commands/review.rb:
```diff
@@ -6,15 +6,38 @@
+        'title' => (global_options[:title]   || GitReflow.current_branch),
+        'body' =>  global_options[:message]
+      }
+      editor = false
```
it still seems strange to have this here... why can't we just change line 22 into an `else`?

codenamev commented on lib/git_reflow/commands/review.rb:
```diff
@@ -6,15 +6,38 @@
 
-    review_options['title'] ||=  GitReflow.get_first_commit_message
-    review_options['body']  ||=  review_options['title']
+    pull_request_msg_file = '.git/GIT_REFLOW_PR_MSG'
```
just remembered there was an PR open with an issue of relative paths... use `git_root_dir = run('git rev-parse --show-toplevel').strip` and prepend it to your path here

codenamev commented on README.rdoc:
```diff
@@ -108,26 +108,35 @@ We assume you know what you're doing, so if you need something different, do it
 After making commits to your branch, run +review+. Didn't push it up? We don't care, we'll do it for you.
 
-  $ git reflow review
+If you do not pass the title or message options to the review command, you will be given an editor to write your PR request in, similar to `git commit`. The first line is the title, the rest is the body. If you pass in the edit option (`-e`) to the review command, it'll open the editor with your branch as the title and last commit as the body, for you to further edit. 
```
cleanup the `-e`s in here

meesterdude commented:
@codenamev updated with your suggestions

codenamev commented:
LGTM

nhance commented:
lgtm

meesterdude commented:
thanks lads!
